lib/rust/enso-prelude/src/data/non_empty_vec.rs

```diff
@@ -0,0 +1,305 @@
+    /// assert_eq!(*vec.last(),2)
+    /// ```
+    pub fn last(&self) -> &T {
+        self.get(self.len() - 1).expect("There is always one element in a NonEmptyVec.")
+    }
+
+    /// Obtain a mutable reference to the last element in the `NonEmptyVec`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use enso_prelude::NonEmptyVec;
+    /// let mut vec = NonEmptyVec::new(0,vec![1,2]);
+    /// assert_eq!(*vec.last_mut(),2)
+    /// ```
+    pub fn last_mut(&mut self) -> &mut T {
+        self.get_mut(self.len() - 1).expect("There is always one element in a NonEmptyVec.")
+    }
+
+    /// Create a draining iterator that removes the specified range in the vector and yields the
+    /// removed items.
+    ///
+    /// It will never remove the root element of the vector.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the starting point is greater than the end point or if the end point is greater
+    /// than the length of the vector.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use enso_prelude::NonEmptyVec;
+    /// let mut vec = NonEmptyVec::new(0,vec![1,2,3,4,5]);
+    /// let drained:Vec<i32> = vec.drain(1..=5).collect();
+    /// assert_eq!(drained,[1,2,3,4,5])
+    /// ```
+    pub fn drain<R>(&mut self, range:R) -> Drain<T> where R:RangeBounds<usize> {
+        if range.contains(&0) {
+            match range.end_bound() {
+                Bound::Included(n) => self.elems.drain(1..=*n),
+                Bound::Excluded(n) => self.elems.drain(1..*n),
+                Bound::Unbounded   => self.elems.drain(1..)
+            }
+        } else {
+            self.elems.drain(range)
+        }
+    }
+
+    /// Creates a splicing iterator that replaces the specified range in the vector with the given 4
+    /// `replace_with` iterator and yields the removed items.
+    ///
+    /// `replace_with` does not need to be the same length as range. The element range is removed
+    /// even if the iterator is not consumed until the end.
+    ///
+    /// It is unspecified how many elements are removed from the vector if the Splice value is leaked.
+    ///
+    /// The input iterator replace_with is only consumed when the Splice value is dropped.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the starting point is greater than the end point or if the end point is greater
+    /// than the length of the vector.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use enso_prelude::NonEmptyVec;
+    /// let mut vec        = NonEmptyVec::new(0,vec![1,2,3,4,5]);
+    /// let replacements   = [10,20,30,40];
+    /// let yielded:Vec<_> = vec.splice(..2,replacements.iter().cloned()).collect();
+    /// assert_eq!(vec.as_slice(),&[10,20,30,40,2,3,4,5]);
+    /// assert_eq!(yielded,&[0,1])
+    /// ```
+    pub fn splice<R,I>(&mut self, range:R, replace_with:I) -> Splice<<I as IntoIterator>::IntoIter>
+    where
+        I: IntoIterator<Item = T>,
+        R: RangeBounds<usize> {
```

Comment:
we're leaving `I : IntoIterator` in the same line as `where` and we align to `I` then. We can ofc talk about changing it, but for now, were doing it this way